### PR TITLE
[WIP] Add flannel support

### PR DESCRIPTION
**Pull request by bogdando**
#### What does this PR do?

* Document flannel options. See the [updated ref arch](https://access.redhat.com/documentation/en-us/reference_architectures/2017/html-single/deploying_and_managing_red_hat_openshift_container_platform_3.6_on_red_hat_openstack_platform_10/#flannel_considerations) for more details.
* Provision a dedicated isolated neutron subnet for the flannel network and disable ports security on it.
* Add openshift-ansible as a galaxy dependency module.
* Create a post-install playbook for Flannel SDN post deployment configs
* Use the os_firewall role to configure iptables and apply rules required for Flannel SDN to function.
* Configure and enable the masked iptables service on the app nodes for Flannel SDN. Keep it as is, disabled and masked, for other nodes groups.

#### How should this be manually tested?
Uncomment flannel options in OSEv3 group vars.
Provision an env, deploy from `openshift-ansible-3.7.0-0.127.0-69-gc16d16d` or later (it must contain the flannel fixes `98db8ec` and `fdc61b8`), and run e2e, it should pass.
Verify that iptables rules added by post-install playbook are persistent across node reboots and iptables services restarts.

#### Is there a relevant Issue open for this?
Provide a link to any open issues that describe the problem you are solving.

#### Who would you like to review this?
cc: @tomassedovic @e-minguez  PTAL

### CODE REVIEW

**Comment by bogdando** [comment]
I can ping the DNS node from a container with an IP address of a given flannel_network. But I still can not access the 53 port of it

**Comment by tomassedovic** [comment]
I'm a bit swamped -- will take a proper look later.

The two things I remember doing in the openshift-on-openstack Heat for flannel are:

We create a second port & subnet for the flannel traffic and for the flannel port we disable port security:

https://github.com/redhat-openstack/openshift-on-openstack/blob/c98db0f69af48acd0b56ba90807398b217d96073/sdn_flannel.yaml#L34

I don't really remember why, but I think the port security bit is actually important. Not sure about using a separate network (but iirc if you disable port security, security groups no longer work so it may be a necessity, too).

**Comment by tomassedovic** [comment]
Thanks for clarification, Edu!

**Comment by bogdando** [comment]
Thanks @tomassedovic @e-minguez ! good for us, we have ansible inventory fixes to deploy multiple private networks in place :) so I'll include the required changes to the heat templates and supporting tasks to disable ports security as well.

**Comment by e-minguez** [comment]
Quick note about this, we saw some issues regarding iptables, so we've updated the documentation with the proper instructions. This needs to be done in the post installation steps.
Check them here -> https://access.redhat.com/documentation/en-us/reference_architectures/2017/html/deploying_red_hat_openshift_container_platform_3.4_on_red_hat_openstack_platform_10/emphasis_manual_deployment_emphasis#run_ansible_installer
Also, for the 3.6 paper (it will be released soon), we have a whole new chapter just for flannel (I'll send you the link once it is published)
Thx!

**Comment by bogdando** [comment]
remaining bits:

* [iptables magic](https://access.redhat.com/documentation/en-us/reference_architectures/2017/html/deploying_red_hat_openshift_container_platform_3.4_on_red_hat_openstack_platform_10/emphasis_manual_deployment_emphasis#run_ansible_installer)
* Update DNS servers hook for the data subnet as well

**Comment by bogdando** [comment]
@e-minguez 

> That's why we use hacks as inserting the rules "live" in the /etc/sysconfig/iptables

yeah, now I can see! thanks! :) There is also a note in the docs, `Do not include Docker or OpenShift Container Platform rules`, but that's exactly what I've attempted to do.

**Comment by e-minguez** [comment]
Just in case, I've created a PR for the official docs: https://github.com/openshift/openshift-docs/pull/5360

**Comment by bogdando** [comment]
@e-minguez with all fixes in place, I have the same issue. The `k8s_POD_cakephp-mysql-example.*build` container has no IP assigned, while mysql pod has it just fine.

**Comment by bogdando** [comment]
basically, I have two pods working with flannel network OK:
```
[
    {
        "Name": "bridge",
        "Id": "902ca5eced42c658fa05a7b70f46b031d49b9333155feb0eb02c26ac40536280",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "10.128.20.1/23",
                    "Gateway": "10.128.20.1"
                }
            ]
        },
        "Internal": false,
        "Containers": {
            "1632bb049973af03d75d138feafd0d7927b7f7b75c5e5b2612ac0aa1062ac339": {
                "Name": "k8s_POD_registry-console-1-bw8tk_default_bc407ed8-a1d0-11e7-a9f0-fa163ebb7eb4_9",
                "EndpointID": "9d3d9ec0a0bb930b49b939bc91cb38ef79400aafa45c10f80f805181387f8cfd",
                "MacAddress": "02:42:0a:80:14:02",
                "IPv4Address": "10.128.20.2/23",
                "IPv6Address": ""
            },
            "ea0cc95e5fde5c2c8e3d86fd14022b682f400c59c9414428fb72bdd077d81606": {
                "Name": "k8s_POD_mysql-1-n1sdc_foo_3686c314-a207-11e7-9711-fa163ebb7eb4_0",
                "EndpointID": "f643207486e212b59627a9f559ec0c28cf0e3ce9764baec82076ccb53a63ed32",
                "MacAddress": "02:42:0a:80:14:03",
                "IPv4Address": "10.128.20.3/23",
                "IPv6Address": ""
            }
        },
        "Options": {
            "com.docker.network.bridge.default_bridge": "true",
            "com.docker.network.bridge.enable_icc": "true",
            "com.docker.network.bridge.enable_ip_masquerade": "true",
            "com.docker.network.bridge.host_binding_ipv4": "0.0.0.0",
            "com.docker.network.bridge.name": "docker0",
            "com.docker.network.driver.mtu": "1450"
        },
        "Labels": {}
    }
]
```

**Comment by bogdando** [comment]
@e-minguez following your tips for getting flannel config from etcd, I have this https://pastebin.com/FvmE98SN PTAL? thanks!

**Comment by bogdando** [comment]
So, looking into the k8s iptables rules, the cake php pod is missing the DNAT rule to translate the service IP to the flannel pod ip. Mysql pod has the rules in place. That should be the root cause, but I have no idea how to debug it further.

```
mysql pod
-A KUBE-SEP-PCVNZH6MZTGCMEJU -s 10.128.20.3/32 -m comment --comment "foo/mysql:mysql" -j KUBE-MARK-MASQ
-A KUBE-SEP-PCVNZH6MZTGCMEJU -p tcp -m comment --comment "foo/mysql:mysql" -m tcp -j DNAT --to-destination 10.128.20.3:3306
-A KUBE-SVC-CXULPK757A3SU6II -m comment --comment "foo/mysql:mysql" -j KUBE-SEP-PCVNZH6MZTGCMEJU
-A KUBE-SERVICES -d 172.30.43.147/32 -p tcp -m comment --comment "foo/mysql:mysql cluster IP" -m tcp --dport 3306 -j KUBE-SVC-CXULPK757A3SU6II

cakephp pod
-A KUBE-SERVICES -d 172.30.114.34/32 -p tcp -m comment --comment "foo/cakephp-mysql-example:web cluster IP" -m tcp --dport 8080 -j KUBE-SVC-7FNEVXMUJVIR6GMX
```
@e-minguez @tomassedovic ^^

**Comment by e-minguez** [comment]
Did you modified the default pod network cidr or services network? There is a PR I did to fix this https://bugzilla.redhat.com/show_bug.cgi?id=1473858

**Comment by bogdando** [comment]
@e-minguez the [net config](https://pastebin.com/FvmE98SN) looks the same as in that [bz](https://bugzilla.redhat.com/show_bug.cgi?id=1473858#c7)

**Comment by e-minguez** [comment]
I've tested it in my environment which has been fully deployed using the 3.6 ref. arch. paper, including iptables rules, etc., and the fixes for pod and services network and:

```
[cloud-user@bastion ~]$ oc get pods -o wide
NAME                               READY     STATUS      RESTARTS   AGE       IP            NODE
cakephp-mysql-persistent-1-build   0/1       Completed   0          6m        10.128.83.2   app-node-1.edu.flannel.com
cakephp-mysql-persistent-1-x650z   1/1       Running     0          3m        10.128.58.3   app-node-0.edu.flannel.com
mysql-1-d76tv                      1/1       Running     0          6m        10.128.83.3   app-node-1.edu.flannel.com
[cloud-user@bastion ~]$ oc get svc
NAME                       CLUSTER-IP       EXTERNAL-IP   PORT(S)    AGE
cakephp-mysql-persistent   10.111.183.84    <none>        8080/TCP   6m
mysql                      10.111.248.204   <none>        3306/TCP   6m
```

Regarding connectivity:

```
[cloud-user@bastion ~]$ oc describe pod mysql-1-d76tv | grep -i port
    Port:		3306/TCP
[cloud-user@bastion ~]$ oc rsh cakephp-mysql-persistent-1-x650z
sh-4.2$ curl -v 10.128.83.3:3306
* About to connect() to 10.128.83.3 port 3306 (#0)
*   Trying 10.128.83.3...
* Connected to 10.128.83.3 (10.128.83.3) port 3306 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 10.128.83.3:3306
> Accept: */*
> 
* Connection #0 to host 10.128.83.3 left intact
Cn���WX>     s:qP7Hmysql_native_password��Got packets out of ordersh-4.2$ ^C
sh-4.2$ curl -v 10.111.248.204:3306
* About to connect() to 10.111.248.204 port 3306 (#0)
*   Trying 10.111.248.204...
* Connected to 10.111.248.204 (10.111.248.204) port 3306 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 10.111.248.204:3306
> Accept: */*
> 
* Connection #0 to host 10.111.248.204 left intact
5.7.16�3r]^���gNyMe*9OJmysql_native_password��Got packets out of ordersh-4.2$ ^C
```
So, the cake php can reach the mysql pod ip and mysql service

```
[cloud-user@bastion ~]$ oc rsh mysql-1-d76tv
sh-4.2$ curl -v 10.128.58.3:8080 | head
* About to connect() to 10.128.58.3 port 8080 (#0)
*   Trying 10.128.58.3...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to 10.128.58.3 (10.128.58.3) port 8080 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 10.128.58.3:8080
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Tue, 26 Sep 2017 08:09:52 GMT
< Server: Apache/2.4.25 (Red Hat)
< Content-Length: 63780
< Content-Type: text/html; charset=UTF-8
< 
{ [data not shown]
<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <title>Welcome to OpenShift</title>


<style>

* Failed writing body (2556 != 16384)
 47 63780   47 30212    0     0  1397k      0 --:--:-- --:--:-- --:--:-- 1475k
* Closing connection 0
curl: (23) Failed writing body (2556 != 16384)
sh-4.2$ curl -v 10.111.183.84:8080 | head
* About to connect() to 10.111.183.84 port 8080 (#0)
*   Trying 10.111.183.84...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to 10.111.183.84 (10.111.183.84) port 8080 (#0)
> GET / HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 10.111.183.84:8080
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Tue, 26 Sep 2017 08:10:26 GMT
< Server: Apache/2.4.25 (Red Hat)
< Content-Length: 63787
< Content-Type: text/html; charset=UTF-8
< 
{ [data not shown]
<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
  <title>Welcome to OpenShift</title>


<style>

* Failed writing body (2556 != 16384)
 47 63787   47 30212    0     0  1393k      0 --:--:-- --:--:-- --:--:-- 1475k
* Closing connection 0
curl: (23) Failed writing body (2556 != 16384)
```

So the mysql pod can reach cake-php pod and service.

And iptables rules here -> http://termbin.com/521m

**Comment by bogdando** [comment]
@e-minguez I found the moby issue looking similar https://github.com/kubernetes/kubernetes/issues/43988
I have the same error message for the cake php pod and the very similar Network hash for the docker network section of it

**Comment by bogdando** [comment]
fwiw, I can manually run pods with flannel networking configured OK `docker run -itd docker.io/openshift/origin-pod`. So the issue is somewhere deep in the k8s/origin and flannel configuration for it.

**Comment by bogdando** [comment]
I split off the [dedicated data network part](https://github.com/openshift/openshift-ansible-contrib/pull/757), that works.

**Comment by bogdando** [comment]
I'll give it another try with `openshift_use_crio=yes` and openshift-ansible-3.7 master

**Comment by bogdando** [comment]
With cri-o enabled, all pods stuck in ContainerCreating then ended up as DeadlineExceeded, so no luck

**Comment by bogdando** [comment]
Created origin issue https://github.com/openshift/origin/issues/16578

**Comment by bogdando** [comment]
Thankfully to @e-minguez this is unblocked now!
For some reason, there is no iptables rules on app hosts to allow containers to reach DNS resolvers.
more details in https://bugzilla.redhat.com/show_bug.cgi?id=1490820

**Comment by bogdando** [comment]
I hope that would be a fix https://github.com/openshift/openshift-ansible/pull/5560

**Comment by bogdando** [comment]
The only remaining issue with persistence of the rules, is the 'docker accept' rule doesn't get restored after a reboot. And I have no idea why :(

That's that I have after a reboot. Compare it the persisted iptables rules before the reboot.
```
[openshift@app-node-0 ~]$ sudo iptables-save | grep -E -e custom -e 'OS_FIREWALL_ALLOW.*\b53'
-A OS_FIREWALL_ALLOW -p tcp -m state --state NEW -m tcp --dport 53 -j ACCEPT
-A OS_FIREWALL_ALLOW -p udp -m state --state NEW -m udp --dport 53 -j ACCEPT
-A POSTROUTING -o eth1 -m comment --comment custom-flannel-docker-2 -j MASQUERADE

[openshift@app-node-0 ~]$ sudo cat /etc/sysconfig/iptables | grep -E -e custom -e 'OS_FIREWALL_ALLOW.*\b53'
-A POSTROUTING -o eth1 -m comment --comment custom-flannel-docker-2 -j MASQUERADE
-A DOCKER -m comment --comment custom-flannel-docker-1 -j ACCEPT
-A OS_FIREWALL_ALLOW -p tcp -m state --state NEW -m tcp --dport 53 -j ACCEPT
-A OS_FIREWALL_ALLOW -p udp -m state --state NEW -m udp --dport 53 -j ACCEPT
```

File: playbooks/provisioning/openstack/post-install.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+- hosts: OSEv3
+  gather_facts: False
+  become: True
+  tasks:
+    - name: Save iptables rules to a backup file
+      when: openshift_use_flannel|default(False)|bool
+      shell: iptables-save > /etc/sysconfig/iptables.orig-$(date +%Y%m%d%H%M%S)
+
+# Enable iptables service on app nodes to persist custom rules (flannel SDN)
+# FIXME(bogdando) w/a https://bugzilla.redhat.com/show_bug.cgi?id=1490820
```
Comment (bogdando) [inline playbooks/provisioning/openstack/post-install.yml]: @e-minguez so with this w/a, it kinda works with the provider, and doesn't spoil the openshift-ansible roles... The persistence issues gone as well, once I used the os_firewall module as a dependency. So the app nodes end up with the iptables service enabled. The other nodes, like masters and infras, will keep running with iptables disabled and masked (just as it is now, I do not change that).

I really have no better way to deal with all of that mess.

**Comment by e-minguez** [comment]
For the persistent rules in the ref. arch. paper, instead creating an iptable rule to allow all protocols traffic in the DOCKER chain, we modify the DOCKER chain to accept everything (:DOCKER ACCEPT):

```
"tac /etc/sysconfig/iptables.orig | sed -e '0,/:DOCKER -/ s/:DOCKER -/:DOCKER ACCEPT/' | awk '"\!"p && /POSTROUTING/{print \"-A POSTROUTING -o eth1 -j MASQUERADE\"; p=1} 1' | tac > /etc/sysconfig/iptables"
```

That was the best way to do it that I found (I'm not proud of the "hack" it just works)

**Comment by bogdando** [comment]
@e-minguez the iptables command here does the same with docker accept. I'd keep it that way cuz on my centos 7.3 env, iptables-restore throws an error `Can't set policy `ACCEPT' on `DOCKER' line X: Bad built-in chain name`

**Comment by bogdando** [comment]
The non flat sec groups case needs to be fixed to allow some things that flannel wants (and that works just fine for the flat sec groups)

**Comment by bogdando** [comment]
I've managed to break this PR so it stopped getting updated with new commits pushed :)
I'll resubmit as a new one.

**Pull request closed without merging**